include/slb/detail/invoke.hpp

```diff
@@ -0,0 +1,384 @@
+            typename PT1 = typename p0704<T, T1>::type,
+            typename Tag = typename dispatch_mem_ptr<C, PT1>::type,
+            typename R =
+                typename mem_fun_ptr_result<Tag, T C::*(PT1&&, Tn&&...)>::type,
+            typename = typename apply_clang_workaround<C, T, T1>::type>
+  constexpr R operator()(T1&& t1, Tn&&... tn) const noexcept(noexcept(
+      _call<R>(Tag{}, _pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...))) {
+    return _call<R>(Tag{}, _pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...);
+  }
+};
+
+} // namespace _invoke
+
+// We don't use `std::is_function<T>` here since some libstdc++ versions fail to
+// classify function types with noexcept specifiers.
+
+template <typename T>
+struct is_function {
+  static constexpr bool value =
+      !std::is_const<T const>::value && !std::is_reference<T>::value;
+};
+
+template <typename T>
+struct mem_fn_result {};
+
+template <typename T, typename C>
+struct mem_fn_result<T C::*> {
+  using type = typename std::conditional<is_function<T>::value,
+                                         _invoke::mem_fun_ptr<T C::*>,
+                                         _invoke::mem_obj_ptr<T C::*>>::type;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+namespace _invoke {
+
+template <typename F, typename FD = typename detail::lib::remove_cvref<F>::type>
+struct dispatch_invoke {
+  using type = F&&;
+};
+
+template <typename F, typename T, typename C>
+struct dispatch_invoke<F, T C::*> {
+  using type = typename mem_fn_result<T C::*>::type;
+};
+
+} // namespace _invoke
+
+template <typename F, typename... Args>
+constexpr auto invoke(F&& f, Args&&... args) noexcept(
+    noexcept(typename _invoke::dispatch_invoke<F>::type(std::forward<F>(f))(
+        std::forward<Args>(args)...)))
+    -> decltype(typename _invoke::dispatch_invoke<F>::type(std::forward<F>(f))(
+        std::forward<Args>(args)...)) {
+  return typename _invoke::dispatch_invoke<F>::type(std::forward<F>(f))(
+      std::forward<Args>(args)...);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+namespace _invoke {
+
+// `INVOKE(f, t1, t2, ..., tN)` implicitly converted to `R`.
+template <typename R, bool IsVoid = std::is_void<R>::value>
+struct invoke_guard {
+  static R _conversion(R) noexcept;
+
+  template <typename F, typename... Args>
+  static constexpr auto call(F&& f, Args&&... args) noexcept(
+      noexcept(_conversion(detail::invoke(std::forward<F>(f),
+                                          std::forward<Args>(args)...))))
+      -> decltype(_conversion(detail::invoke(std::forward<F>(f),
+                                             std::forward<Args>(args)...))) {
+    return detail::invoke(std::forward<F>(f), std::forward<Args>(args)...);
+  }
+};
+
+// `static_cast<void>(INVOKE(f, t1, t2, ..., tN))` if `R` is _cv_ `void`.
+template <typename R>
+struct invoke_guard<R, /*IsVoid:*/ true> {
```

Comment:
Hm, why not leave out the `bool IsVoid` parameter and just specialize on `void` like this?
```
template <>
struct invoke_guard<void> { ... };
```

Reply:
There are four void types: plain `void` and the cv-qualified versions.

Reply:
Would it be worth turning this into

```
template <typename R, typename RD = std::remove_cv_t<R>>
struct invoke_guard;

template <typename R>
struct invoke_guard<R, void> { /*...*/ };
```

instead?

Reply:
Ah~ it's because `is_void` handles the `cv`-qualifiers. I see.
That's a bit subtle, but I think the `bool` is fine.

Reply:
The explicit removal of cv-qualifiers should make it non-subtle. I'll make the changes.